### PR TITLE
feat: evaluate project using grammar rules

### PR DESCRIPTION
**Pull request by cmatsuoka**
Grammar evaluation is done for a specific set of build-on/build-for
architectures, however those are listed in the build plan which in
turn is obtained from the project model. To keep the grammar
handling transparent to the application we evaluate grammar twice,
first to obtain the build plan, and then to resolve for the correct set
of architectures.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [ ] Have you followed the guidelines for contributing?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox`?

-----

(CRAFT-2322)

### CODE REVIEW

**Comment by sergiusens** [comment]
we should live review this one

File: craft_application/application.py [hunk 1]
```diff
@@ -188,9 +187,20 @@ def project(self) -> models.Project:
         with project_file.open() as file:
             yaml_data = util.safe_yaml_load(file)
 
+        # validate project grammar
+        GrammarAwareProject.validate_grammar(yaml_data)
+
         yaml_data = self._transform_project_yaml(yaml_data)
 
-        return self.app.ProjectClass.from_yaml_data(yaml_data, project_file)
+        if "parts" in yaml_data:
```
Comment (lengau) [inline craft_application/application.py:195]: Let's see if we can put this in process_project_yaml

File: craft_application/application.py [hunk 2]
```diff
@@ -188,9 +187,20 @@ def project(self) -> models.Project:
...
+            craft_cli.emit.debug(f"Processing grammar (on {build_on} for {build_for})")
+            yaml_data["parts"] = grammar.process_parts(
```
Comment (cmatsuoka) [inline craft_application/application.py:197]: note: this only modify part

File: craft_application/application.py [hunk 3]
```diff
@@ -178,8 +178,7 @@ def _configure_services(
             work_dir=self._work_dir,
         )
 
-    @functools.cached_property
-    def project(self) -> models.Project:
+    def load_project(self, build_on: str, build_for: str) -> None:
```
Comment (lengau) [inline craft_application/application.py:181]: We have a moving target here, as while you were implementing this I changed `project` to this:
https://github.com/canonical/craft-application/blob/f584223edd3e56f78f59a7f4a0895a55ed1d117e/craft_application/application.py#L211

Comment (cmatsuoka) [reply craft_application/application.py]: This should be easy to reconcile once we settle on a grammar integration strategy.

**Comment by cmatsuoka** [comment]
@lengau Can you check if this implementation satisfies charmcraft requirements?

**Comment by lengau** [comment]
@cmatsuoka I think this will work fine for Charmcraft. The only potential concern I have is related to actually implementing grammar in charmcraft (which it currently isn't), that we're fully parsing the grammar before we do the "magic" parts bits in charmcraft. But that's most likely just a limitation in what we can do with grammar, and it's in something that Charmcraft doesn't even implement yet.

File: tests/conftest.py [hunk 4]
```diff
@@ -31,7 +31,16 @@
     from collections.abc import Iterator
 
 
-class MyProject(models.Project):
+class Platform(models.CraftBaseModel):
```
Comment (lengau) [inline tests/conftest.py:34]: Not strictly related to this, but do we have a full spec of what the platforms section will look like?

File: craft_application/models/project.py [hunk 5]
```diff
@@ -52,6 +52,25 @@ class BuildInfo:
...
 
 
+class BuildPlannerConfig(CraftBaseConfig):
+    """Config for BuildProjects."""
+
+    extra = pydantic.Extra.ignore
+    """The BuildPlanner model uses attributes from the project yaml."""
+
+
+class BuildPlanner(CraftBaseModel):
```
Comment (lengau) [inline craft_application/models/project.py:62]: I like moving this off the project, but before the 2.0 release I'd like to discuss where to actually put the `get_build_plan` method. Out of scope here, but it's something that doesn't quite feel like it has an obvious home.

**Review by lengau — CHANGES_REQUESTED** [review]
1.5 requested changes 😄 

1. Using the same function for the alias generator (I'm ok with making it public for that purpose)
2. Grammar as a disable-able feature. This is the half change because I'm very persuadable with a "that's a bad idea" response too.

File: craft_application/application.py [hunk 6]
```diff
@@ -454,6 +487,15 @@ def _transform_project_yaml(self, yaml_data: dict[str, Any]) -> dict[str, Any]:
         if self.app.features.build_secrets:
             self._render_secrets(yaml_data)
 
+        # Expand grammar.
+        if build_for and "parts" in yaml_data:
```
[the remark this answers is not in the document]
Comment (lengau) [reply craft_application/application.py:491]: (Based on direct conversation we're making this a standard part of craft-application)

File: tests/unit/test_grammar.py
Comment (lengau) [inline tests/unit/test_grammar.py:1]: These tests are great!

**Review by lengau — APPROVED** [review]
Looks great, thank you!

File: craft_application/models/project.py [hunk 7]
```diff
@@ -64,7 +83,9 @@ class Project(CraftBaseModel):
     summary: Optional[SummaryStr]
     description: Optional[str]
 
-    base: Optional[Any]
+    base: Optional[Any] = None
+    build_base: Optional[Any] = None
```
Comment (tigarmo) [inline craft_application/models/project.py:87]: where are these new fields used?

File: craft_application/models/project.py [hunk 8]
```diff
@@ -55,6 +55,25 @@ class BuildInfo:
...
+    """The BuildPlanner obtains a build plan for the project."""
+
+    Config = BuildPlannerConfig
+
+    def get_build_plan(self) -> List[BuildInfo]:
+        """Obtain the list of architectures and bases from the project file."""
+        raise NotImplementedError(
+            f"{self.__class__.__name__!s} must implement get_build_plan"
+        )
+
```
[the remark this answers is not in the document]
Comment (tigarmo) [reply craft_application/models/project.py:75]: do we still need to keep `Project.get_build_plan()`?

[the remark this answers is not in the document]
Comment (tigarmo) [reply craft_application/models/project.py:75]: but I don't think any other code calls `get_build_plan()`

File: craft_application/models/project.py (hunk 7 above)
Comment (syu-w) [reply craft_application/models/project.py:87]: I was thinking they are used in the `craft-grammar`, but seems not? @cmatsuoka

**Review by tigarmo — APPROVED** [review]

File: craft_application/models/project.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (syu-w) [reply craft_application/models/project.py:75]: Yes, the craft-application itself does not use it for now. However others may use it, but I did not check if that is true or actually nobody using it

**Pull request merged**